Add OpenGrok to Sourcegraph search syntax guide (#11384)
sourcegraph/sourcegraph: doc/user/search/opengrok.md

## Related pages

- [All Sourcegraph search documentation](index.md)
  - [Sourcegraph query syntax](queries.md)
- [Search product feature comparisons](https://about.sourcegraph.com/workflow)
- [OpenGrok to Sourcegraph admin migration guide](../../admin/migration/opengrok.md)

## Switching to Sourcegraph's query syntax

### Wildcards vs. regular expressions

Oracle OpenGrok provides wildcard support for searches. For example, to find all strings beginning with `foo`, you can use the wildcard search `foo*`. Similarly, OpenGrok provides the `?` operator for single character wildcards.

Sourcegraph, [provides full regular expression search](queries.md#regexp-search), with support for the [RE2 syntax](https://golang.org/s/re2syntax). The same search above would take the form `foo.*` (or in this case, just `foo`, since Sourcegraph automatically supports partial matches). Much more powerful regexp expressions are available.

> (Note that Sourcegraph also provides a [literal search mode](queries.md#Literal-search--default-) by default, in which there's no need to escape special characters. This simplifies searches such a `foo(`, which would result in an error in regexp mode.)

Reviewer: Nit: this is only true for patterns in literal search mode. I.e., there is still a need to escape special characters in filters. I'm not saying we need to clarify here, but just to give careful thought if you think a casual reader may get that impression from the current description.

Author: That makes sense, but the start of this sentence says:

> Note that Sourcegraph also provides a [literal search mode](queries.md#Literal-search--default-) by default, in which

Is that not sufficient?

Reviewer: yeah think it's fine :-)

Sourcegraph also provides the ability to search on multiple Git revisions in a single repository at once, using `:` characters to separate revision names in the `repo:` field. For example, search both `feature-branch` and `abc123` in a single query using `pattern repo:github.com/org/repo@feature-branch:abc123`.

### Special characters

Oracle OpenGrok doesn't index most single-character strings (such as for special characters like `{`, `}`, `[`, `]`, `+`, `-`, and more), and non-alpha-numeric characters generally.

Sourcegraph indexes all characters, and can search for strings of any length. Using the default [literal search mode](queries.md#Literal-search--default-), any search (including those with special characters like `foo.bar`, `try {`, `i++`, `i-=1`, `foo->bar`, and more), will all be searchable without special handling. Using [regexp mode](queries.md#regexp-search) would require escaping special charactes.

> The only exceptions are colon characters, which are by default used for specifying a search keyword on Sourcegraph (see below). Any search containing colons can be done using the `content:"pattern"` keyword to explicitly mark it as the search string.

Reviewer: ```suggestion
The only exceptions are colon characters, which are by default used for specifying a search keyword on Sourcegraph (see below). Any search containing colons can be done using, for example, `content:"contains:colon"` to explicitly mark it as the search string.
```
Not strongly attached to this suggestion, just a thought.

Author: ```suggestion
The only exceptions are colon characters, which are by default used for specifying a search keyword on Sourcegraph (see below). Any search containing colons can be done using, for example, `content:"foo::bar"` keyword to explicitly mark it as the search string.
```

I see what you are suggesting here, but I'm afraid it only gets more complicated as we make it recursive 😄 . This is a realistic example that helps

### Boolean operators

Oracle OpenGrok provides three boolean operators — `AND`, `OR`, and `NOT` — for scoping searches to files that contain strings that match multiple patterns.

Sourcegraph provides [`AND` and `OR` operators](queries.md#operators).

> NOTE: Operators are available as of Sourcegraph 3.15 and enabled with `{"experimentalFeatures": {"andOrQuery": "enabled"}}` in the site configuration. Built-in operator support is planned for an upcoming release.

### Search keywords

Both Sourcegraph and OpenGrok allow users to add keywords for scoping searches. Below is a mapping from OpenGrok syntax to Sourcegraph.

| | OpenGrok | Sourcegraph |
|-------------------|------------|--------|
| Search text | `full:pattern` | `pattern` |
| Search symbol definitions | `def:pattern` | `pattern type:symbol` |
| Search symbol references | `def:pattern` | Available through hover tooltips and `Find references` panels on code pages |
| Search for repository names | Not supported | `pattern type:repo` |

> | Search for file names | `file:pattern` | `pattern type:repo` |

Reviewer: Is that a typo for the Sourcegraph column?, should it be `type:path`?
```suggestion
| Search for file names | `file:pattern` | `pattern type:path` |
```
Also, I much prefer we encourage `file:pattern` rather than `type:` because the values for `type:` are not great (`"file"` actually means content). Though, I am aware that using `file:pattern` may not be exactly the same, do you have a concrete example why we couldn't just say `file:pattern` here?

Author: Ah, nope, that sounds good to me! I didn't realize that, actually... Should we change the UI to make that clear?

Author: ```suggestion
| Search for file names | `file:pattern` | `file:pattern` |
```

> | Search commit messages | `hist:pattern` | `pattern type:path` |

Reviewer: ```suggestion
| Search commit messages | `hist:pattern` | `pattern type:commit` |
```

| Search code changes (diff search) | Not supported | `pattern type:diff` |
| Scope searches to a language | `pattern type:c` | `pattern lang:c` |
| Case sensitivity | Not supported | `case:yes` or `case:no` |
| Scope searches to forked repositories | Supported through the repository selector | `fork:yes`, `fork:no`, or `fork:only` |
| Scope searches to archived repositories | Supported through the repository selector | `archived:yes`, `archived:no`, or `archived:only` |
| Scope searches to repositories that contain a file | Not supported | `pattern repohasfile:foo` |
| Scope searches to recently updated repositories | Not supported | `pattern repohascommitafter:"3 months"` or `pattern repohascommitafter:"june 25 2017"` |

Sourcegraph also provides keywords to [scope commit message and diff searches](queries.md#keywords--diff-and-commit-searches-only-) to specific authors or timeframes in which a change was made.

To see an exhaustive list of Sourcegraph's search keywords, see the [search query syntax](queries.md#keywords--all-searches-) page.

**Negating search keywords**

> Sourcegraph allows users to negate any search keyword by appending a `-` character to the name. For example, `pattern -repo:foo` will search across all repositories except for those that contain "foo". `pattern -lang:javascript` will search across all non-JavaScript files.

Reviewer: We may have to mention here or in the table above that we don't support negating content search yet. I think in OpenGrok this is done with something like `-"foo"`, but I'm not sure (I know this is how it is in [Lucene](https://lucene.apache.org/core/2_9_4/queryparsersyntax.html))

Author: ```suggestion
Oracle OpenGrok allows users to negate any search keyword, or to negate file results that contain a given search pattern, by appending a `-` character to the start of the term.

Sourcegraph also allows users to negate search keywords by appending a `-` character to the name. For example, `pattern -repo:foo` will search across all repositories except for those that contain "foo". `pattern -lang:javascript` will search across all non-JavaScript files. Sourcegraph cannot currently exclude files based on text that they don't contain, but this will be available using [operators](queries.md#operators) soon.
```

Author: Good call! Despite poor docs on this, I just tested and it looks like OpenGrok DOES support negation of all terms. I've updated this now to reflect that.
